codegen/src/test/resources/software/amazon/awssdk/codegen/poet/paginators/PaginatedOperationWithResultKeyPublisher.java

```diff
@@ -0,0 +1,149 @@
+ * operation which is a paginated operation. This class is a type of {@link org.reactivestreams.Publisher} which can be
+ * used to provide a sequence of
+ * {@link software.amazon.awssdk.services.jsonprotocoltests.model.PaginatedOperationWithResultKeyResponse} response
+ * pages as per demand from the subscriber.
+ * </p>
+ * <p>
+ * When the operation is called, an instance of this class is returned. At this point, no service calls are made yet and
+ * so there is no guarantee that the request is valid. If there are errors in your request, you will see the failures
+ * only after you start streaming the data. The subscribe method should be called as a request to start streaming data.
+ * For more info, see {@link org.reactivestreams.Publisher#subscribe(org.reactivestreams.Subscriber)}. Each call to the
+ * subscribe method will result in a new {@link org.reactivestreams.Subscription} i.e., a new contract to stream data
+ * from the starting request.
+ * </p>
+ *
+ * <p>
+ * The following are few ways to use the response class:
+ * </p>
+ * 1) Using the forEach helper method
+ *
+ * <pre>
+ * {@code
+ * software.amazon.awssdk.services.jsonprotocoltests.paginators.PaginatedOperationWithResultKeyPublisher publisher = client.paginatedOperationWithResultKeyPaginator(request);
+ * CompletableFuture<Void> future = publisher.forEach(res -> { // Do something with the response });
+ * future.get();
+ * }
+ * </pre>
+ *
+ * 2) Using a custom subscriber
+ *
+ * <pre>
+ * {@code
+ * software.amazon.awssdk.services.jsonprotocoltests.paginators.PaginatedOperationWithResultKeyPublisher publisher = client.paginatedOperationWithResultKeyPaginator(request);
+ * publisher.subscribe(new Subscriber<software.amazon.awssdk.services.jsonprotocoltests.model.PaginatedOperationWithResultKeyResponse>() {
+ *
+ * public void onSubscribe(org.reactivestreams.Subscriber subscription) { //... };
+ *
+ *
+ * public void onNext(software.amazon.awssdk.services.jsonprotocoltests.model.PaginatedOperationWithResultKeyResponse response) { //... };
+ * });}
+ * </pre>
+ *
+ * As the response is a publisher, it can work well with third party reactive streams implementations like RxJava2.
+ * <p>
+ * <b>Note: If you prefer to have control on service calls, use the
+ * {@link #paginatedOperationWithResultKey(software.amazon.awssdk.services.jsonprotocoltests.model.PaginatedOperationWithResultKeyRequest)}
+ * operation.</b>
+ * </p>
+ */
+@Generated("software.amazon.awssdk:codegen")
+public class PaginatedOperationWithResultKeyPublisher implements SdkPublisher<PaginatedOperationWithResultKeyResponse> {
+    private final JsonProtocolTestsAsyncClient client;
+
+    private final PaginatedOperationWithResultKeyRequest firstRequest;
+
+    private final AsyncPageFetcher nextPageFetcher;
+
+    private boolean isLastPage;
```

Comment:
What's this used for?

Reply:
In resume method, we return EmptySubscription (EmptyIterator for sync) on subscribe if there are no more pages. This works when customer subscribes to get pages. If you try to subscribe to get Items (PaginatedItemsPublisher class), there is currently no way to recognize if input is last page and return EmptySubscription. The two options I see are:
1) Pass a value to PaginatedItemsPublisher to indicate its last page (current approach)
2) Override all the items methods in resume method to use EmptySubscription (if there are no more pages). I felt this will bloat all the publisher classes and chose option 1. 
The resume method in option 2 will look like:

```
    public PaginatedOperationWithResultKeyPublisher resume(final PaginatedOperationWithResultKeyResponse lastSuccessfulPage) {
        if (nextPageFetcher.hasNextPage(lastSuccessfulPage)) {
            return new PaginatedOperationWithResultKeyPublisher(client, firstRequest.toBuilder()
                                                                                    .nextToken(lastSuccessfulPage.nextToken()).build());
        }
        return new PaginatedOperationWithResultKeyPublisher(client, firstRequest) {
            @Override
            public void subscribe(Subscriber<? super PaginatedOperationWithResultKeyResponse> subscriber) {
                subscriber.onSubscribe(new EmptySubscription(subscriber));
            }

            @Override
            public SdkPublisher<SimpleStruct> items() {
                return new EmptyPublisher<>(); // A new publisher that creates EmptySubscription when subscribed to
            }
            
            //More overriden methods if there are multiple result keys
        };
    }
```